tyrian/js/src/main/scala/tyrian/websocket/WebSocket.scala

```diff
@@ -102,59 +107,56 @@ object WebSocket:
       onOpenSendMessage: Option[String],
       settings: KeepAliveSettings
   ): F[LiveSocket[F]] =
-    Async[F].delay {
-      val socket    = new dom.WebSocket(address)
-      val keepAlive = new KeepAlive(socket, settings)
-      val q         = Queue.empty[WebSocketEvent]
-
-      val msgListener = Functions.fun { e =>
-        val event = WebSocketEvent.Receive(e.asInstanceOf[dom.MessageEvent].data.toString)
-        q.enqueue(event)
-      }
-
-      val errListener = Functions.fun { e =>
-        val msg =
-          try e.asInstanceOf[dom.ErrorEvent].message
-          catch { case _: Throwable => "Unknown" }
-        q.enqueue(WebSocketEvent.Error(msg))
-      }
-
-      val closeListener = Functions.fun { e =>
-        val ev = e.asInstanceOf[dom.CloseEvent]
-        q.enqueue(WebSocketEvent.Close(ev.code, ev.reason))
-      }
-
-      val openListener = Functions.fun { _ =>
-        onOpenSendMessage.foreach(msg => socket.send(msg))
-        q.enqueue(WebSocketEvent.Open)
-      }
-
-      socket.addEventListener("message", msgListener)
-      socket.addEventListener("error", errListener)
-      socket.addEventListener("open", openListener)
-      socket.addEventListener("close", closeListener)
-
-      val subs =
-        Sub.Batch(
-          Sub.make(s"[tyrian-ws-${address}]")(
-            Stream.repeatEval {
-              Sync[F].delay {
-                try Some(q.dequeue())
-                catch _ => None
-              }
-            }.unNone
-          )(
-            Async[F].delay {
-              socket.removeEventListener("message", msgListener)
-              socket.removeEventListener("error", errListener)
-              socket.removeEventListener("open", openListener)
-              socket.removeEventListener("close", closeListener)
-            }
-          ),
-          keepAlive.run
-        )
-
-      LiveSocket(socket, subs)
+    (Channel.unbounded[F, WebSocketEvent], Dispatcher.sequential[F].allocated).flatMapN {
+      case (channel, (dispatcher, closeDispatcher)) =>
+        Async[F].delay {
+          val socket    = new dom.WebSocket(address)
+          val keepAlive = new KeepAlive(socket, settings)
+
+          val msgListener = Functions.fun { e =>
+            val event = WebSocketEvent.Receive(e.asInstanceOf[dom.MessageEvent].data.toString)
+            dispatcher.unsafeRunAndForget(channel.send(event))
+          }
+
+          val errListener = Functions.fun { e =>
+            val msg =
+              try e.asInstanceOf[dom.ErrorEvent].message
+              catch { case _: Throwable => "Unknown" }
+            dispatcher.unsafeRunAndForget(channel.send(WebSocketEvent.Error(msg)))
+          }
+
+          val closeListener = Functions.fun { e =>
+            val ev = e.asInstanceOf[dom.CloseEvent]
+            dispatcher.unsafeRunAndForget(channel.send(WebSocketEvent.Close(ev.code, ev.reason)))
+          }
```

Comment:
So Gabriel raises a really important issue to make sure that the `Dispatcher` and `Channel` are closed on network error. Since errors are propagated through this callback I think this is where we want to add additional logic. Probably same for the close event listener below as well.

@davesmith00000 would it be safe to assume that once the websocket has errored it is no longer usable?

Reply:
Ah yeah, that seems to be exactly the case.

> The error event is fired when a connection with a WebSocket has been closed due to an error (some data couldn't be sent for example).

https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/error_event

Reply:
Sorry I'm late. I know not a lot about WebSockets*, but I read that page the same way you do and agree that closing on error sounds sensible (Which I note you've done in another commit). :+1: 

(*WebSockets have had a lot of attention, but largely but other people! Who knew they'd be so popular?!)